### PR TITLE
Don't include below-dust inbound HTLCs in commit tx fee calculation

### LINKED ISSUE #757: chanmon_consistency found force-close
The test `39ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff343235302a2a2a2a2a2a34343434343434343434343434343434343434fffe` on top of the changes in #753 manages to find a force-close where node 1 sends an HTLC which node 2 considers bogus as "Remote HTLC add would put them under remote reserve value". We (I mean Val) should fix this.

### PR DESCRIPTION
**Pull request by valentinewallace**
Closes #757 

Does this fix look right? Seems like an oops from the original channel reserve PR. Let me know and I'll write tests.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
This looks good to me. I'll let the fuzzer spin on this + 756 for a bit and see what it thinks.

**Comment by TheBlueMatt** [comment]
Fuzzers didn't find anything after 15 CPU-days, so probably at least fixes that failing case. Obviously can't rely on fuzzers so should add an explicit test (yay `lightning/src/ln/chanmon_update_fail_tests.rs`), but otherwise looks good.

**Comment by valentinewallace** [comment]
Still needs a bit more testing but I pushed up one test and an additional fix (the fix being:  when calculating the commit tx fee, don't include the HTLC currently being added if it's below dust).

File: lightning/src/ln/channel.rs [hunk 1]
```diff
@@ -1729,10 +1729,23 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
...
 	// to add a number of additional HTLCs to the calculation. Note that dust
 	// HTLCs are excluded.
-	fn next_local_commit_tx_fee_msat(&self, addl_htlcs: usize) -> u64 {
+	fn next_local_commit_tx_fee_msat(&self, new_htlc_amount: u64, fee_spike_buffer_htlc: Option<()>) -> u64 {
 		assert!(self.channel_outbound);
 
-		let mut their_acked_htlcs = self.pending_inbound_htlcs.len();
+		let mut addl_htlcs = 1;
+		if fee_spike_buffer_htlc.is_some() { addl_htlcs += 1; }
+		if new_htlc_amount / 1000 <= self.counterparty_dust_limit_satoshis {
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs:1737]: Should the counterparty/holder dust limit references in this PR be swapped? If its easy, would be nice to get a test which would catch such an inversion.

**Comment by valentinewallace** [comment]
Found some more missing pieces so I'm restructuring to reuse some of the logic in `build_commitment_transaction`. Will update when this is ready for review again.

**Comment by TheBlueMatt** [comment]
If you're refactoring bits of `build_commitment_transaction` you may want to build on/anticipate #742.

**Comment by TheBlueMatt** [comment]
Is there a way we can do a more thorough test of this? eg in `cfg(test)` store the latest value here and when we call `build_commitment_transaction` check that the latest value matches? It may require a bunch of "do we need to check here" cases, but if its not too bad it may be worth it.

**Comment by valentinewallace** [comment]
> Is there a way we can do a more thorough test of this? eg in `cfg(test)` store the latest value here and when we call `build_commitment_transaction` check that the latest value matches? It may require a bunch of "do we need to check here" cases, but if its not too bad it may be worth it.

That makes sense, currently trying to implement this!

**Comment by valentinewallace** [comment]
> Is there a way we can do a more thorough test of this? eg in `cfg(test)` store the latest value here and when we call `build_commitment_transaction` check that the latest value matches? It may require a bunch of "do we need to check here" cases, but if its not too bad it may be worth it.

Ok, I believe I've approximately added this testing.

**Comment by TheBlueMatt** [comment]
We should probably enable the new fee caching stuff in fuzztarget, too.

**Comment by TheBlueMatt** [comment]
Note that the new tests, with just the `cfg` changed to run in fuzztarget, fail on at least `2c090d2c2c2c340c13fffe039f42f3e7`.

**Comment by TheBlueMatt** [comment]
Ah, I take that back, the `full_stack_target` failure definitely looks like a bug either in the new tests or in the new logic.

**Comment by valentinewallace** [comment]
> Note that the new tests, with just the `cfg` changed to run in fuzztarget, fail on at least `2c090d2c2c2c340c13fffe039f42f3e7`.

OK I believe I've fixed this and `full_stack_target`

**Comment by TheBlueMatt** [comment]
It looks like `2a441414192a441414190a0a2a441414190d61c9de5c021d` is failing chanmon_consistency.

**Comment by valentinewallace** [comment]
> It looks like `2a441414192a441414190a0a2a441414190d61c9de5c021d` is failing chanmon_consistency.

I think my current approach in 076137d9d3f13fe184485f384023bab98f5f8154 causes too many bespoke exceptions to be necessary. Gonna try to rework it.

**Comment by TheBlueMatt** [comment]
Right, I was kinda afraid of that. I'm mostly OK with a ton of exceptions in #[cfg(test)] blocks, but if you can find a simpler way that would be cool too.

File: lightning/src/ln/channel.rs [hunk 2]
```diff
@@ -1684,63 +1714,155 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
...
-			// We only include outbound HTLCs if it will not be included in their next
-			// commitment_signed, i.e. if they've responded to us with an RAA after announcement.
+			// We only include outbound HTLCs if it will not be included in their next commitment_signed,
+			// i.e. if they've responded to us with an RAA after announcement.
 			match htlc.state {
-				OutboundHTLCState::Committed => their_acked_htlcs += 1,
-				OutboundHTLCState::RemoteRemoved {..} => their_acked_htlcs += 1,
+				OutboundHTLCState::Committed => included_htlcs += 1,
+				OutboundHTLCState::RemoteRemoved {..} => included_htlcs += 1,
+				OutboundHTLCState::LocalAnnounced { .. } => included_htlcs += 1,
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:1837]: @TheBlueMatt I added this line because we *do* include the HTLC currently being added. So why would we not include other `LocalAnnounced` HTLCs?

Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:1837]: I think I'm missing context for this question, but I dont think `LocalAnnounced` HTLCs get included in the next remote commitment tx, only the next local commitment tx (and only probably the next remote commitment tx, but only if they give us an RAA first).

Comment (valentinewallace) [reply lightning/src/ln/channel.rs:1837]: > I think I'm missing context for this question,

Lol sorry, I just wanted to draw your attention to the fact that I was adding this line to make sure it makes sense. I'm a bit confused by the rest of your comment, but I think you're saying it *does* make sense to include `LocalAnnounced` HTLCs here.

**Comment by valentinewallace** [comment]
> Right, I was kinda afraid of that. I'm mostly OK with a ton of exceptions in #[cfg(test)] blocks, but if you can find a simpler way that would be cool too.

Lmk what you think of the new approach.

File: lightning/src/ln/channel.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:1837]: Oh, right, yes, I believe so. I had confused myself as to which commitment tx was remove vs local....for the millionth time.

**Comment by TheBlueMatt** [comment]
> Lmk what you think of the new approach.

Right, much better, thanks! I am a little worried about the "info.num_htlcs == counterparty_commitment_tx.1" checks neutering some of the value - isn't part of the reason for the checks to make sure we have the same number of HTLCs as we thought we would? Is there a way to take a bit of both approaches and maybe just drop that check and then reset the states on RAA or commitment_signed messages?

**Comment by valentinewallace** [comment]
> > Lmk what you think of the new approach.
> 
> Right, much better, thanks! I am a little worried about the "info.num_htlcs == counterparty_commitment_tx.1" checks neutering some of the value - isn't part of the reason for the checks to make sure we have the same number of HTLCs as we thought we would? Is there a way to take a bit of both approaches and maybe just drop that check and then reset the states on RAA or commitment_signed messages?

Adding those specific checks^ still had test failures. I agree the `info.num_htlcs == counterparty_commitment_tx.1` did neuter some of the value though, so I've updated it to instead make sure the total number of pending HTLCs is the same. I *think* this should capture the full value of the check.

File: lightning/src/ln/channel.rs [hunk 3]
```diff
@@ -1684,63 +1714,154 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
...
-	fn next_local_commit_tx_fee_msat(&self, addl_htlcs: usize) -> u64 {
+	// Get the commitment tx fee for the local's (i.e. our) next commitment transaction based on the
+	// number of pending HTLCs that are on track to be in our next commitment tx, plus an additional
+	// HTLC if `fee_spike_buffer_htlc` is Some, plus a new HTLC given by `new_htlc_amount`. Dust HTLCs
+	// are excluded.
+	fn next_local_commit_tx_fee_msat(&self, new_htlc_amount: u64, locally_offered: bool, fee_spike_buffer_htlc: Option<()>) -> u64 {
 		assert!(self.is_outbound());
 
-		let mut their_acked_htlcs = self.pending_inbound_htlcs.len();
+		let real_dust_limit_success_sat = (self.feerate_per_kw as u64 * HTLC_SUCCESS_TX_WEIGHT / 1000) + self.holder_dust_limit_satoshis;
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs:1724]: It would be nice to have tests that fail if the HTLC_TIMEOUT/_SUCCESS multipliers are swapped (here and in the next fn). At least for this one, my existing fuzz corpus didn't hit it :(.

**Review by TheBlueMatt — APPROVED** [review]
Right. Looks good to me! I added one note about some manual mutation testing that didn't get caught, if its easy would be nice, but if its a pain I'm ok without.

File: lightning/src/ln/channel.rs (hunk 3 above)
Comment (valentinewallace) [reply lightning/src/ln/channel.rs:1724]: Good idea, added this test!

File: lightning/src/ln/channel.rs [hunk 4]
```diff
@@ -1684,63 +1684,108 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
...
-	// based on the number of pending HTLCs that are on track to be in our next
-	// commitment tx. `addl_htcs` is an optional parameter allowing the caller
-	// to add a number of additional HTLCs to the calculation. Note that dust
-	// HTLCs are excluded.
-	fn next_local_commit_tx_fee_msat(&self, addl_htlcs: usize) -> u64 {
+	// Get the commitment tx fee for the local's (i.e. our) next commitment transaction based on the
+	// number of pending HTLCs that are on track to be in our next commitment tx, plus an additional
+	// HTLC if `fee_spike_buffer_htlc` is Some, plus a new HTLC given by `new_htlc_amount`. Dust HTLCs
+	// are excluded.
+	fn next_local_commit_tx_fee_msat(&self, new_htlc_amount: u64, locally_offered: bool, fee_spike_buffer_htlc: Option<()>) -> u64 {
```
Comment (jkczyz) [inline lightning/src/ln/channel.rs:1721]: I'm not a huge fan of passing bools to a function as it isn't obvious at the call site as to what the argument means. Consider using an enum, possibly combining it with the amount from the first parameter such that the combined parameter represents the HTLC with some value.

File: lightning/src/ln/channel.rs [hunk 5]
```diff
@@ -2019,15 +2095,31 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
...
+				let projected_commit_tx_info = self.next_local_commitment_tx_fee_info_cached.lock().unwrap().take();
+				*self.next_remote_commitment_tx_fee_info_cached.lock().unwrap() = None;
+				if let Some(info) = projected_commit_tx_info {
+					let total_pending_htlcs = self.pending_inbound_htlcs.len() + self.pending_outbound_htlcs.len()
+						+ self.holding_cell_htlc_updates.len();
+					if info.total_pending_htlcs == total_pending_htlcs
+						&& info.next_holder_htlc_id == self.next_holder_htlc_id
+						&& info.next_counterparty_htlc_id == self.next_counterparty_htlc_id
+						&& info.feerate == self.feerate_per_kw {
+							assert_eq!(total_fee, info.fee / 1000);
```
Comment (jkczyz) [inline lightning/src/ln/channel.rs:2118]: Is there any way to write a test that checks that these agree instead of adding to the production code (albeit, only conditionally for tests). Generally, writing a test would be preferable if it is possible.

Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:2118]: I think the point of this block is more to check it in fuzz/functional tests than one-off. Would be great to get the test code our of `channel.rs`, I agree, but not sure how to do that in the fuzzer/functional tests.

File: lightning/src/ln/channel.rs (hunk 4 above)
Comment (valentinewallace) [reply lightning/src/ln/channel.rs:1721]: Sure, would you be fine with:
```
enum HTLCInitiator {
    LocallyOffered
    RemoteOffered
}
struct HTLCCandidate {
    amount_msat: u64,
    origin: HTLCInitiator
}
fn next_local_commit_tx_fee_msat(&self, htlc: HTLCCandidate....) { ... }
```

Comment (jkczyz) [reply lightning/src/ln/channel.rs:1721]: Maybe `LocalOffered` for consistency, not withstanding the glossary's usage of holder/counterparty (or perhaps operator/counterparty in the future). Probably best to stick to the terminology already in use here.

If call sites seem verbose, consider adding `local` and `remote` constructors, or something similarly named.

Comment (valentinewallace) [reply lightning/src/ln/channel.rs]: I think that turned out much nicer, lmk what you think.

Comment (jkczyz) [reply lightning/src/ln/channel.rs]: Looks great! Thanks or making the change.

File: lightning/src/ln/channel.rs [hunk 6]
```diff
@@ -4575,6 +4615,116 @@ mod tests {
...
+		let node_b_node_id = PublicKey::from_secret_key(&secp_ctx, &SecretKey::from_slice(&[7; 32]).unwrap());
+		let node_b_chan = Channel::<EnforcingChannelKeys>::new_from_req(&&feeest, &&keys_provider, node_b_node_id, InitFeatures::known(), &open_channel_msg, 7, &config).unwrap();
+
+		// Node B --> Node A: accept channel, explicitly setting B's dust limit.
+		let mut accept_channel_msg = node_b_chan.get_accept_channel();
+		accept_channel_msg.dust_limit_satoshis = 546;
+		node_a_chan.accept_channel(&accept_channel_msg, &config, InitFeatures::known()).unwrap();
+
+		// Put some inbound and outbound HTLCs in A's channel.
+		let htlc_amount_msat = 11_092_000; // put an amount below A's effective dust amount but above B's.
```
Comment (jkczyz) [inline lightning/src/ln/channel.rs:4814]: nit: s/dust amount/dust limit

**Review by jkczyz — APPROVED** [review]

File: lightning/src/ln/channel.rs [hunk 7]
```diff
@@ -1684,63 +1735,172 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
...
+		for ref htlc in self.pending_inbound_htlcs.iter() {
+			if htlc.amount_msat / 1000 <= real_dust_limit_timeout_sat {
+				continue
+			}
+			included_htlcs += 1;
+		}
+
 		for ref htlc in self.pending_outbound_htlcs.iter() {
-			if htlc.amount_msat / 1000 <= self.counterparty_dust_limit_satoshis {
+			if htlc.amount_msat / 1000 <= real_dust_limit_success_sat {
```
Comment (ariard) [inline lightning/src/ln/channel.rs:1866]: I think it would be great to comment this subtlety around which dust limit apply in function of which viewpoint the fee is computed for.

"When computing the fee as paid by counterparty, if they have to claim onchain HTLC by broadcasting their commitment, they will have to spend inbound through a HTLC-timeout, outbound through a HTLC-Success. 
When computing the fee as paid by holder, if we have to claim onchain HTLC by broadcasting our commitment, we will have to spend inbound through a HTLC-success, outbound through a HTLC-timeout"

Well, that's an attempt, but we should make it clear that even if our channel variables represent the holder direction-side (`pending_inbound_htlcs`/`pending_outbound_htlcs`) the actual fee cost is function from which commitment they will be claim.

**Review by ariard — COMMENTED** [review]
Post-merge Code Review ACK 1079753

Overall, a call for better comments given how much off-chain computation is tricky. Accounting a HTLC for fee computation is function of its state, and even if we follow the most sound reasoning IMO, I wouldn't be surprised that other implementation have weird recipes ("What if they have another definition of `AwaitingRemoteRevokeToRemove` ?"), thus leading to unexpected closure in case of edge-cases. Good documentation would help track down that kind of case.

I think I've also found few test coverage holes.

**Comment by valentinewallace** [comment]
Thanks for the post-merge review @ariard, will attempt to address soon^tm
